indexing-hadoop/src/main/java/io/druid/indexer/hadoop/DatasourceRecordReader.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.hadoop.mapreduce.RecordReader;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+public class DatasourceRecordReader extends RecordReader<NullWritable, HadoopWritableInputRow>
+{
+
+  private static final Logger logger = LoggerFactory.getLogger(DatasourceRecordReader.class);
+
+  private SegmentLoadSpec spec;
+  private Iterable<Metric> complexMetrics;
+
+  private IngestSegmentFirehose rowYielder;
+
+  private MapBasedRow currRow;
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException
+  {
+    spec = readSegmentJobSpec(context.getConfiguration(), new ObjectMapper());
+
+    final List<String> dimensions = spec.getDimensions();
+    final List<String> metrics = Lists.transform(
+        spec.getMetrics(),
+        new Function<Metric, String>()
+        {
+          @Nullable
+          @Override
+          public String apply(Metric input)
+          {
+            return input.getName();
+          }
+        }
+    );
+
+    complexMetrics = Iterables.filter(
+        spec.getMetrics(), new Predicate<Metric>()
+        {
+          @Override
+          public boolean apply(@Nullable Metric input)
+          {
+            return input.getType() == ValueType.COMPLEX;
+          }
+        }
+    );
+    final DimFilter filter = spec.getFilter();
+
+    String hdfsPath = ((DatasourceInputSplit) split).getPath();
+    logger.info("Reading segment from " + hdfsPath);
+
+    File segmentDir = Files.createTempDir();
+    logger.info("segment dir: " + segmentDir);
+
+    FileSystem fs = FileSystem.get(context.getConfiguration());
+    getSegmentFiles(hdfsPath, segmentDir, fs);
+    logger.info("finished getting segment files");
+
+    QueryableIndex index = IndexIO.loadIndex(segmentDir);
+    StorageAdapter adapter = new QueryableIndexStorageAdapter(index);
+    List<StorageAdapter> adapters = Lists.newArrayList(adapter);
+    rowYielder = new IngestSegmentFirehose(adapters, dimensions, metrics, filter, spec.getInterval(), spec.getGranularity());
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException, InterruptedException
+  {
+    if (rowYielder.hasMore()) {
+      currRow = (MapBasedRow)rowYielder.nextRow();
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  @Override
+  public NullWritable getCurrentKey() throws IOException, InterruptedException
+  {
+    return NullWritable.get();
+  }
+
+  @Override
+  public HadoopWritableInputRow getCurrentValue() throws IOException, InterruptedException
+  {
+    Map<String, Object> event = currRow.getEvent();
+    for(Metric m : complexMetrics) {
+      if(event.containsKey(m.getName())) {
+        ComplexMetricSerde serde = ComplexMetrics.getSerdeForType(m.getComplexTypeName());
```

Comment:
this is only converting complex metric "object" to byte[] so that MapBasedHadoopWritableInputRow could have generic serialization code. MapBasedInputRow is not enough as it does not have any serializability contract on it. For hadoop indexing, input rows have to be serializable so that we can pass them to reducers.